### PR TITLE
Support method level labels

### PR DESCRIPTION
**Pull request by divtosz**
Added method level labels and improved form field labels

No additional dependencies required.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Refactoring (non-breaking code restructuring that preserves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build-related update (CI workflow, test cases)
- [ ] Documentation update
- [ ] Other (please describe):

## How Has This Been Tested?

I have updated test/ui/heuristic_labels_test.py to check method level labels as well. This can be run from the main directory as nosetests test/ui/heuristic_labels_test.py

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tkltest/generate/ui/generate_with_labels_analysis.py [hunk 1]
```diff
@@ -31,7 +31,9 @@
 
 # names and paths for generated code files
 _POM_FILE = 'pom.xml'
-_CRAWL_PATHS_FILE = 'crawl_paths_addressbook_small.json'
+
+_CRAWL_PATHS_FILE = 'crawl_paths_tmf.json'
```
Comment (rachelt44) [inline tkltest/generate/ui/generate_with_labels_analysis.py:35]: @divtosz why do specific app files/name appear in the code?

File: tkltest/generate/ui/generate_with_labels_analysis.py [hunk 2]
```diff
@@ -49,7 +51,10 @@ def generate_selenium_api_tests(config, crawl_dir):
         crawl_dir (dict): Root crawl directory created by Crawljax for the current test-generation run
     """
     logging.info('Creating Selenium API tests from paths in {}'.format(os.path.join(crawl_dir, _CRAWL_PATHS_FILE)))
-    app_name = config['general']['app_name']
+
+    # app_name = config['general']['app_name']
+    app_name = _CRAWL_PATHS_FILE[12:-5]
```
Comment (rachelt44) [inline tkltest/generate/ui/generate_with_labels_analysis.py:56]: @divtosz can the hard-coded values be replaced with constants?

File: tkltest/generate/ui/generate_with_labels_analysis.py [hunk 3]
```diff
@@ -108,23 +116,29 @@ def generate_selenium_api_tests(config, crawl_dir):
...
-        logging.info('Path {}: length={}, {}'.format(path_num, len(crawl_path), method_name))
-        if method_name in method_name_count:
-            method_name_count[method_name] = method_name_count[method_name] + 1
-            method_name = '{}_dup{}'.format(method_name, method_name_count[method_name])
-            logging.info('Creating duplicate test method: {}'.format(method_name))
+        method_eventables_path = __create_method_name_for_path(crawl_path)
+        logging.info('Path {}: length={}, {}'.format(path_num, len(crawl_path), method_eventables_path))
+        if method_eventables_path in method_eventables_path_count:
+            method_eventables_path_count[method_eventables_path] = method_eventables_path_count[method_eventables_path] + 1
+            method_eventables_path = '{}_dup{}'.format(method_eventables_path, method_eventables_path_count[method_eventables_path])
```
Comment (rachelt44) [inline tkltest/generate/ui/generate_with_labels_analysis.py:129]: @divtosz should we indeed create a duplicate method, or skip the duplicate path?

File: tkltest/generate/ui/heuristic_labels.py [hunk 4]
```diff
@@ -183,19 +281,90 @@ def get_label(self, eventable):
...
+            form_field_extended_dom (str/dom): extended dom with label attribute for the form field
+
+        """
+        form_field_dom = self.find_element(source_dom, form_field_xpath)
+        if form_field_dom is None:
+            return None
+        i = 0
+        while form_field_dom.getparent():
+            form_field_dom = form_field_dom.getparent()
+            if i == 2 or form_field_dom.get('label') is not None or form_field_dom.get('aria-label') is not None:
```
Comment (rachelt44) [inline tkltest/generate/ui/heuristic_labels.py:316]: @divtosz document what is special in i ==2

**Review by rachelt44 — CHANGES_REQUESTED** [review]

File: tkltest/generate/ui/generate_with_labels_analysis.py (hunk 1 above)
Comment (divtosz) [reply tkltest/generate/ui/generate_with_labels_analysis.py:35]: this is just for testing purposes

File: tkltest/generate/ui/generate_with_labels_analysis.py (hunk 2 above)
Comment (divtosz) [reply tkltest/generate/ui/generate_with_labels_analysis.py:56]: This is also for testing purposes, following the file format crawl_paths_<webappname>.json, just so files don't get overwritten; before the output files were getting overwritten.

File: tkltest/generate/ui/heuristic_labels.py [hunk 5]
```diff
@@ -183,19 +281,90 @@ def get_label(self, eventable):
...
+        Get label for all the form fields for this eventable element
+
+        Parameters:
+            eventable (dict): A json dict of the eventable
+
+        Returns:
+            form_field_labels (list): a list of labels corresponding to all the form fields for this eventable
+
+        """
+        form_field_labels = []
```
[the remark this answers is not in the document]
Comment (divtosz) [reply tkltest/generate/ui/heuristic_labels.py:335]: I will do this and re-push

File: tkltest/generate/ui/generate_with_labels_analysis.py (hunk 3 above)
Comment (sinha108) [reply tkltest/generate/ui/generate_with_labels_analysis.py:129]: @divtosz Why is the code being copied? Shouldn't the loop over tests in CrawlPaths.json be implemented once?

Comment (divtosz) [reply tkltest/generate/ui/generate_with_labels_analysis.py:129]: No I wrote this file to run the test generation with a static file, without crawljax creating the paths, just for analysis purposes and to make testing faster

**Review by sinha108 — APPROVED** [review]

**Pull request merged**